File: snippets/screens/swiftui/explore_screen/3/explore_screen.swift
import SwiftUI

struct ExploreScreen: View {
    @StateObject private var viewModel = ExploreViewModel()
    @State private var searchText = ""
    @State private var selectedTab = 0

    var body: some View {
        NavigationView {
            VStack(spacing: 0) {
                // Search bar
                SearchBar(text: $searchText, onSubmit: {
                    viewModel.performSearch(searchText)
                })
                .padding(.horizontal, 16)
                .padding(.vertical, 12)

                // Tab picker
                Picker("Explore", selection: $selectedTab) {
                    Text("Trending").tag(0)
                    Text("Tags").tag(1)
                    Text("People").tag(2)
                    Text("Places").tag(3)
                }
                .pickerStyle(.segmented)
                .padding(.horizontal, 16)
                .padding(.bottom, 12)

                // Content
                TabView(selection: $selectedTab) {
                    TrendingView(viewModel: viewModel)
                        .tag(0)

                    TagsView(viewModel: viewModel)
                        .tag(1)

                    PeopleView(viewModel: viewModel)
                        .tag(2)

                    PlacesView(viewModel: viewModel)
                        .tag(3)
                }
                .tabViewStyle(.page(indexDisplayMode: .never))
            }
            .navigationTitle("Explore")
            .navigationBarTitleDisplayMode(.large)
        }
        .onAppear {
            viewModel.loadExploreContent()
        }
    }
}

struct TrendingView: View {
    @ObservedObject var viewModel: ExploreViewModel

    let columns = Array(repeating: GridItem(.flexible(), spacing: 2), count: 3)

    var body: some View {
        ScrollView {
            VStack(spacing: 12) {
                // Category filters
                CategoryFilterView(
                    selectedCategory: $viewModel.selectedCategory,
                    categories: viewModel.categories
                )

                // Featured trending section
                FeaturedTrendingCard()
                    .padding(.horizontal, 16)

                // Posts grid
                LazyVGrid(columns: columns, spacing: 2) {
                    ForEach(viewModel.filteredPosts) { post in
                        TrendingPostCard(post: post) {
                            viewModel.viewPost(post)
                        }
                    }
                }
                .padding(.horizontal, 2)
            }
        }
        .refreshable {
            await viewModel.refreshTrending()
        }
    }
}

struct CategoryFilterView: View {
    @Binding var selectedCategory: String
    let categories: [String]

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 12) {
                ForEach(categories, id: \.self) { category in
                    FilterChip(
                        title: category,
                        isSelected: selectedCategory == category,
                        onTap: {
                            selectedCategory = category
                        }
                    )
                }
            }
            .padding(.horizontal, 16)
        }
    }
}

struct FeaturedTrendingCard: View {
    var body: some View {
        RoundedRectangle(cornerRadius: 12)
            .fill(
                LinearGradient(
                    colors: [.purple, .blue],
                    startPoint: .topLeading,
                    endPoint: .bottomTrailing
                )
            )
            .frame(height: 120)
            .overlay(
                VStack {
                    Image(systemName: "chart.line.uptrend.xyaxis")
                        .font(.system(size: 32))
                        .foregroundColor(.white)

                    Text("Trending Now")
                        .font(.headline)
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                }
            )
    }
}

struct TrendingPostCard: View {
    let post: ExplorePost
    let onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            ZStack {
                AsyncImage(url: URL(string: post.imageUrl)) { image in
                    image
                        .resizable()
                        .aspectRatio(1, contentMode: .fill)
                } placeholder: {
                    Rectangle()
                        .fill(Color.gray.opacity(0.3))
                        .overlay(
                            Image(systemName: "photo")
                                .foregroundColor(.gray)
                        )
                }
                .clipped()

                // Video indicator
                if post.isVideo {
                    VStack {
                        HStack {
                            Spacer()
                            HStack(spacing: 2) {
                                Image(systemName: "play.fill")
                                    .font(.caption2)
                                Text(post.duration ?? "")
                                    .font(.caption2)
                            }
                            .foregroundColor(.white)
                            .padding(.horizontal, 6)
                            .padding(.vertical, 2)
                            .background(Color.black.opacity(0.6))
                            .cornerRadius(4)
                            .padding(8)
                        }
                        Spacer()
                    }
                }

                // Like count overlay
                VStack {
                    Spacer()
                    HStack {
                        HStack(spacing: 4) {
                            Image(systemName: "heart.fill")
                                .font(.caption)
                            Text(formatCount(post.likes))
                                .font(.caption)
                        }
                        .foregroundColor(.white)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(Color.black.opacity(0.6))
                        .cornerRadius(4)
                        .padding(8)

                        Spacer()
                    }
                }
            }
        }
        .buttonStyle(.plain)
    }

    private func formatCount(_ count: Int) -> String {
        if count < 1000 { return "\(count)" }
        if count < 1000000 { return String(format: "%.1fK", Double(count) / 1000) }
        return String(format: "%.1fM", Double(count) / 1000000)
    }
}

struct TagsView: View {
    @ObservedObject var viewModel: ExploreViewModel

    let columns = Array(repeating: GridItem(.flexible(), spacing: 16), count: 2)

    var body: some View {
        ScrollView {
            LazyVGrid(columns: columns, spacing: 16) {
                ForEach(viewModel.trendingTags, id: \.self) { tag in
                    TagCard(tag: tag) {
                        viewModel.searchByTag(tag)
                    }
                }
            }
            .padding(.horizontal, 16)
        }
    }
}

struct TagCard: View {
    let tag: String
    let onTap: () -> Void

    private var postCount: Int {
        abs(tag.hashCode) % 1000000
    }

    private var gradientColors: [Color] {
        switch tag.lowercased() {
        case "#trending":
            return [.orange, .red]
        case "#photography":
            return [.blue, .purple]
        case "#travel":
            return [.teal, .blue]
        case "#food":
            return [.orange, .red]
        case "#fashion":
            return [.pink, .purple]
        case "#fitness":
            return [.green, .teal]
        case "#art":
            return [.purple, .pink]
        case "#nature":
            return [.green, .mint]
        default:
            return [.gray, .secondary]
        }
    }

    var body: some View {
        Button(action: onTap) {
            RoundedRectangle(cornerRadius: 12)
                .fill(
                    LinearGradient(
                        colors: gradientColors,
                        startPoint: .topLeading,
                        endPoint: .bottomTrailing
                    )
                )
                .frame(height: 120)
                .overlay(
                    VStack(alignment: .leading) {
                        Text(tag)
                            .font(.title2)
                            .fontWeight(.bold)
                            .foregroundColor(.white)

                        Spacer()

                        Text("\(formatCount(postCount)) posts")
                            .font(.subheadline)
                            .foregroundColor(.white.opacity(0.8))
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(16)
                )
        }
        .buttonStyle(.plain)
    }

    private func formatCount(_ count: Int) -> String {
        if count < 1000 { return "\(count)" }
        if count < 1000000 { return String(format: "%.1fK", Double(count) / 1000) }
        return String(format: "%.1fM", Double(count) / 1000000)
    }
}

struct PeopleView: View {
    @ObservedObject var viewModel: ExploreViewModel

    var body: some View {
        ScrollView {
            LazyVStack(spacing: 16) {
                ForEach(viewModel.suggestedUsers) { user in
                    SuggestedUserCard(user: user) {
                        viewModel.followUser(user)
                    }
                }
            }
            .padding(.horizontal, 16)
        }
    }
}

struct SuggestedUserCard: View {
    let user: SuggestedUser
    let onFollow: () -> Void

    var body: some View {
        HStack(spacing: 16) {
            AsyncImage(url: URL(string: user.avatar)) { image in
                image
                    .resizable()
                    .aspectRatio(contentMode: .fill)
            } placeholder: {
                Circle()
                    .fill(Color.gray.opacity(0.3))
            }
            .frame(width: 60, height: 60)
            .clipShape(Circle())

            VStack(alignment: .leading, spacing: 4) {
                HStack {
                    Text(user.username)
                        .font(.headline)
                        .fontWeight(.semibold)

                    if user.isVerified {
                        Image(systemName: "checkmark.seal.fill")
                            .font(.caption)
                            .foregroundColor(.blue)
                    }
                }

                Text(user.displayName)
                    .font(.subheadline)
                    .foregroundColor(.secondary)

                Text("\(formatCount(user.followersCount)) followers")
                    .font(.caption)
                    .foregroundColor(.secondary)
            }

            Spacer()

            Button("Follow") {
                onFollow()
            }
            .font(.subheadline)
            .fontWeight(.semibold)
            .foregroundColor(.white)
            .padding(.horizontal, 20)
            .padding(.vertical, 8)
            .background(Color.blue)
            .cornerRadius(8)
        }
        .padding(.all, 16)
        .background(Color(UIColor.secondarySystemBackground))
        .cornerRadius(12)
    }

    private func formatCount(_ count: Int) -> String {
        if count < 1000 { return "\(count)" }
        if count < 1000000 { return String(format: "%.1fK", Double(count) / 1000) }
        return String(format: "%.1fM", Double(count) / 1000000)
    }
}

struct PlacesView: View {
    @ObservedObject var viewModel: ExploreViewModel

    var body: some View {
        ScrollView {
            LazyVStack(spacing: 16) {
                ForEach(viewModel.places, id: \.name) { place in
                    PlaceCard(place: place) {
                        viewModel.searchByPlace(place.name)
                    }
                }
            }
            .padding(.horizontal, 16)
        }
    }
}

struct PlaceCard: View {
    let place: Place
    let onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            ZStack {
                AsyncImage(url: URL(string: place.image)) { image in
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                } placeholder: {
                    Rectangle()
                        .fill(Color.gray.opacity(0.3))
                }
                .frame(height: 120)
                .clipped()
                .cornerRadius(12)

                // Gradient overlay
                LinearGradient(
                    colors: [.clear, .black.opacity(0.6)],
                    startPoint: .top,
                    endPoint: .bottom
                )
                .cornerRadius(12)

                // Text overlay
                VStack(alignment: .leading) {
                    Spacer()
                    HStack {
                        VStack(alignment: .leading, spacing: 2) {
                            Text(place.name)
                                .font(.headline)
                                .fontWeight(.bold)
                                .foregroundColor(.white)

                            Text(place.posts)
                                .font(.subheadline)
                                .foregroundColor(.white.opacity(0.8))
                        }
                        Spacer()
                    }
                    .padding(16)
                }
            }
        }
        .buttonStyle(.plain)
    }
}

@MainActor
class ExploreViewModel: ObservableObject {
    @Published var posts: [ExplorePost] = []
    @Published var filteredPosts: [ExplorePost] = []
    @Published var selectedCategory = "All"
    @Published var trendingTags: [String] = []
    @Published var suggestedUsers: [SuggestedUser] = []
    @Published var places: [Place] = []

    let categories = ["All", "Photos", "Videos", "People", "Tags"]

    func loadExploreContent() {
        loadPosts()
        loadTags()
        loadUsers()
        loadPlaces()
    }

    private func loadPosts() {
        posts = [
            ExplorePost(
                id: "1",
                imageUrl: "https://example.com/trending1.jpg",
                likes: 2456,
                comments: 123,
                username: "travel_photographer",
                isVideo: false,
                tags: ["travel", "landscape", "sunset"]
            ),
            ExplorePost(
                id: "2",
                imageUrl: "https://example.com/trending2.jpg",
                likes: 1834,
                comments: 89,
                username: "food_lover",
                isVideo: true,
                duration: "0:45",
                tags: ["food", "recipe", "cooking"]
            ),
            ExplorePost(
                id: "3",
                imageUrl: "https://example.com/trending3.jpg",
                likes: 3021,
                comments: 234,
                username: "street_artist",
                isVideo: false,
                tags: ["art", "street", "graffiti"]
            )
        ]

        filteredPosts = posts
    }

    private func loadTags() {
        trendingTags = [
            "#trending",
            "#photography",
            "#travel",
            "#food",
            "#fashion",
            "#fitness",
            "#art",
            "#nature",
            "#lifestyle",
            "#music",
            "#dance",
            "#comedy"
        ]
    }

    private func loadUsers() {
        suggestedUsers = [
            SuggestedUser(
                username: "amazing_photographer",
                displayName: "Alex Chen",
                avatar: "https://example.com/suggested1.jpg",
                isVerified: true,
                followersCount: 125000
            ),
            SuggestedUser(
                username: "chef_master",
                displayName: "Maria Rodriguez",
                avatar: "https://example.com/suggested2.jpg",
                isVerified: false,
                followersCount: 85000
            ),
            SuggestedUser(
                username: "fitness_coach",
                displayName: "John Smith",
                avatar: "https://example.com/suggested3.jpg",
                isVerified: true,
                followersCount: 200000
            )
        ]
    }

    private func loadPlaces() {
        places = [
            Place(
                name: "New York City",
                posts: "2.3M posts",
                image: "https://example.com/nyc.jpg"
            ),
            Place(
                name: "Tokyo",
                posts: "1.8M posts",
                image: "https://example.com/tokyo.jpg"
            ),
            Place(
                name: "Paris",
                posts: "1.5M posts",
                image: "https://example.com/paris.jpg"
            ),
            Place(
                name: "London",
                posts: "1.2M posts",
                image: "https://example.com/london.jpg"
            )
        ]
    }

    func performSearch(_ query: String) {
        // Navigate to search results
    }

    func searchByTag(_ tag: String) {
        // Navigate to search results for tag
    }

    func searchByPlace(_ place: String) {
        // Navigate to search results for place
    }

    func viewPost(_ post: ExplorePost) {
        // Navigate to post detail
    }

    func followUser(_ user: SuggestedUser) {
        // Follow user
    }

    func refreshTrending() async {
        // Simulate API call
        try? await Task.sleep(nanoseconds: 1_000_000_000)
        posts.shuffle()
        filteredPosts = posts
    }
}

struct ExplorePost: Identifiable {
    let id: String
    let imageUrl: String
    let likes: Int
    let comments: Int
    let username: String
    let isVideo: Bool
    let duration: String?
    let tags: [String]

    init(id: String, imageUrl: String, likes: Int, comments: Int, username: String, isVideo: Bool, duration: String? = nil, tags: [String]) {
        self.id = id
        self.imageUrl = imageUrl
        self.likes = likes
        self.comments = comments
        self.username = username
        self.isVideo = isVideo
        self.duration = duration
        self.tags = tags
    }
}

struct SuggestedUser: Identifiable {
    let id = UUID()
    let username: String
    let displayName: String
    let avatar: String
    let isVerified: Bool
    let followersCount: Int
}

struct Place {
    let name: String
    let posts: String
    let image: String
}

extension String {
    var hashCode: Int {
        var hasher = Hasher()
        hasher.combine(self)
        return hasher.finalize()
    }
}

#Preview {
    ExploreScreen()
}

File: snippets/screens/flutter/explore_screen/2/explore_screen.dart
import 'package:flutter/material.dart';

class ExploreScreen extends StatefulWidget {
  @override
  _ExploreScreenState createState() => _ExploreScreenState();
}

class _ExploreScreenState extends State<ExploreScreen>
    with SingleTickerProviderStateMixin {
  final TextEditingController _searchController = TextEditingController();
  late TabController _tabController;

  List<Post> trendingPosts = [
    Post(
      id: '1',
      imageUrl: 'https://example.com/trending1.jpg',
      likes: 2456,
      comments: 123,
      username: 'travel_photographer',
      isVideo: false,
      tags: ['travel', 'landscape', 'sunset'],
    ),
    Post(
      id: '2',
      imageUrl: 'https://example.com/trending2.jpg',
      likes: 1834,
      comments: 89,
      username: 'food_lover',
      isVideo: true,
      duration: '0:45',
      tags: ['food', 'recipe', 'cooking'],
    ),
    Post(
      id: '3',
      imageUrl: 'https://example.com/trending3.jpg',
      likes: 3021,
      comments: 234,
      username: 'street_artist',
      isVideo: false,
      tags: ['art', 'street', 'graffiti'],
    ),
    Post(
      id: '4',
      imageUrl: 'https://example.com/trending4.jpg',
      likes: 987,
      comments: 67,
      username: 'fitness_guru',
      isVideo: true,
      duration: '1:20',
      tags: ['fitness', 'workout', 'health'],
    ),
    Post(
      id: '5',
      imageUrl: 'https://example.com/trending5.jpg',
      likes: 1567,
      comments: 98,
      username: 'fashion_icon',
      isVideo: false,
      tags: ['fashion', 'style', 'outfit'],
    ),
    Post(
      id: '6',
      imageUrl: 'https://example.com/trending6.jpg',
      likes: 2198,
      comments: 156,
      username: 'nature_lover',
      isVideo: false,
      tags: ['nature', 'wildlife', 'photography'],
    ),
  ];

  List<String> trendingTags = [
    '#trending',
    '#photography',
    '#travel',
    '#food',
    '#fashion',
    '#fitness',
    '#art',
    '#nature',
    '#lifestyle',
    '#music',
    '#dance',
    '#comedy',
  ];

  List<User> suggestedUsers = [
    User(
      username: 'amazing_photographer',
      displayName: 'Alex Chen',
      avatar: 'https://example.com/suggested1.jpg',
      isVerified: true,
      followersCount: 125000,
    ),
    User(
      username: 'chef_master',
      displayName: 'Maria Rodriguez',
      avatar: 'https://example.com/suggested2.jpg',
      isVerified: false,
      followersCount: 85000,
    ),
    User(
      username: 'fitness_coach',
      displayName: 'John Smith',
      avatar: 'https://example.com/suggested3.jpg',
      isVerified: true,
      followersCount: 200000,
    ),
  ];

  String selectedCategory = 'All';
  List<String> categories = ['All', 'Photos', 'Videos', 'People', 'Tags'];

  @override
  void initState() {
    super.initState();
    _tabController = TabController(length: 4, vsync: this);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(
          'Explore',
          style: TextStyle(fontWeight: FontWeight.bold),
        ),
        backgroundColor: Colors.white,
        foregroundColor: Colors.black,
        elevation: 1,
        bottom: PreferredSize(
          preferredSize: Size.fromHeight(120),
          child: Column(
            children: [
              // Search bar
              Padding(
                padding: EdgeInsets.all(16),
                child: TextField(
                  controller: _searchController,
                  decoration: InputDecoration(
                    hintText: 'Search photos, videos, and people...',
                    prefixIcon: Icon(Icons.search),
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(25),
                      borderSide: BorderSide.none,
                    ),
                    filled: true,
                    fillColor: Colors.grey[100],
                    contentPadding: EdgeInsets.symmetric(horizontal: 16, vertical: 8),
                  ),
                  onSubmitted: _performSearch,
                ),
              ),
              // Tab bar
              TabBar(
                controller: _tabController,
                labelColor: Colors.black,
                unselectedLabelColor: Colors.grey,
                indicatorColor: Colors.black,
                isScrollable: true,
                tabs: [
                  Tab(text: 'Trending'),
                  Tab(text: 'Tags'),
                  Tab(text: 'People'),
                  Tab(text: 'Places'),
                ],
              ),
            ],
          ),
        ),
      ),
      body: TabBarView(
        controller: _tabController,
        children: [
          _buildTrendingTab(),
          _buildTagsTab(),
          _buildPeopleTab(),
          _buildPlacesTab(),
        ],
      ),
    );
  }

  Widget _buildTrendingTab() {
    return RefreshIndicator(
      onRefresh: _refreshTrending,
      child: CustomScrollView(
        slivers: [
          // Category filters
          SliverToBoxAdapter(
            child: Container(
              height: 50,
              child: ListView.builder(
                scrollDirection: Axis.horizontal,
                padding: EdgeInsets.symmetric(horizontal: 16),
                itemCount: categories.length,
                itemBuilder: (context, index) {
                  return _buildCategoryChip(categories[index]);
                },
              ),
            ),
          ),
          // Posts grid
          SliverPadding(
            padding: EdgeInsets.all(2),
            sliver: SliverGrid(
              gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                crossAxisCount: 3,
                crossAxisSpacing: 2,
                mainAxisSpacing: 2,
                childAspectRatio: 1.0,
              ),
              delegate: SliverChildBuilderDelegate(
                (context, index) {
                  if (index == 0) {
                    return _buildFeaturedPost();
                  }
                  final postIndex = (index - 1) % trendingPosts.length;
                  return _buildPostTile(trendingPosts[postIndex]);
                },
                childCount: trendingPosts.length + 1,
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildTagsTab() {
    return GridView.builder(
      padding: EdgeInsets.all(16),
      gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
        crossAxisCount: 2,
        crossAxisSpacing: 16,
        mainAxisSpacing: 16,
        childAspectRatio: 1.5,
      ),
      itemCount: trendingTags.length,
      itemBuilder: (context, index) {
        return _buildTagCard(trendingTags[index]);
      },
    );
  }

  Widget _buildPeopleTab() {
    return ListView.separated(
      padding: EdgeInsets.all(16),
      itemCount: suggestedUsers.length,
      separatorBuilder: (context, index) => SizedBox(height: 16),
      itemBuilder: (context, index) {
        return _buildUserCard(suggestedUsers[index]);
      },
    );
  }

  Widget _buildPlacesTab() {
    final places = [
      {'name': 'New York City', 'posts': '2.3M posts', 'image': 'https://example.com/nyc.jpg'},
      {'name': 'Tokyo', 'posts': '1.8M posts', 'image': 'https://example.com/tokyo.jpg'},
      {'name': 'Paris', 'posts': '1.5M posts', 'image': 'https://example.com/paris.jpg'},
      {'name': 'London', 'posts': '1.2M posts', 'image': 'https://example.com/london.jpg'},
    ];

    return ListView.builder(
      padding: EdgeInsets.all(16),
      itemCount: places.length,
      itemBuilder: (context, index) {
        return _buildPlaceCard(places[index]);
      },
    );
  }

  Widget _buildCategoryChip(String category) {
    bool isSelected = selectedCategory == category;
    return Padding(
      padding: EdgeInsets.only(right: 8),
      child: FilterChip(
        label: Text(
          category,
          style: TextStyle(
            color: isSelected ? Colors.white : Colors.grey[700],
          ),
        ),
        selected: isSelected,
        onSelected: (selected) {
          setState(() {
            selectedCategory = category;
          });
        },
        selectedColor: Theme.of(context).primaryColor,
        backgroundColor: Colors.grey[100],
        side: BorderSide.none,
      ),
    );
  }

  Widget _buildFeaturedPost() {
    return GestureDetector(
      onTap: () => _viewPost(trendingPosts.first),
      child: Container(
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(8),
          gradient: LinearGradient(
            colors: [Colors.purple, Colors.blue],
            begin: Alignment.topLeft,
            end: Alignment.bottomRight,
          ),
        ),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Icon(Icons.trending_up, color: Colors.white, size: 32),
            SizedBox(height: 8),
            Text(
              'Trending',
              style: TextStyle(
                color: Colors.white,
                fontWeight: FontWeight.bold,
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildPostTile(Post post) {
    return GestureDetector(
      onTap: () => _viewPost(post),
      child: Stack(
        children: [
          Container(
            width: double.infinity,
            height: double.infinity,
            child: Image.network(
              post.imageUrl,
              fit: BoxFit.cover,
              errorBuilder: (context, error, stackTrace) {
                return Container(
                  color: Colors.grey[300],
                  child: Icon(Icons.image, color: Colors.grey),
                );
              },
            ),
          ),
          if (post.isVideo)
            Positioned(
              top: 8,
              right: 8,
              child: Container(
                padding: EdgeInsets.symmetric(horizontal: 6, vertical: 2),
                decoration: BoxDecoration(
                  color: Colors.black54,
                  borderRadius: BorderRadius.circular(4),
                ),
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Icon(Icons.play_arrow, color: Colors.white, size: 12),
                    Text(
                      post.duration ?? '',
                      style: TextStyle(color: Colors.white, fontSize: 10),
                    ),
                  ],
                ),
              ),
            ),
          Positioned(
            bottom: 0,
            left: 0,
            right: 0,
            child: Container(
              padding: EdgeInsets.all(8),
              decoration: BoxDecoration(
                gradient: LinearGradient(
                  begin: Alignment.topCenter,
                  end: Alignment.bottomCenter,
                  colors: [Colors.transparent, Colors.black54],
                ),
              ),
              child: Row(
                children: [
                  Icon(Icons.favorite, color: Colors.white, size: 16),
                  SizedBox(width: 4),
                  Text(
                    _formatCount(post.likes),
                    style: TextStyle(color: Colors.white, fontSize: 12),
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildTagCard(String tag) {
    final postCount = (tag.hashCode % 1000000).abs();

    return GestureDetector(
      onTap: () => _searchByTag(tag),
      child: Container(
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(12),
          gradient: LinearGradient(
            colors: _getTagGradient(tag),
            begin: Alignment.topLeft,
            end: Alignment.bottomRight,
          ),
        ),
        padding: EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              tag,
              style: TextStyle(
                color: Colors.white,
                fontSize: 18,
                fontWeight: FontWeight.bold,
              ),
            ),
            Spacer(),
            Text(
              '${_formatCount(postCount)} posts',
              style: TextStyle(
                color: Colors.white70,
                fontSize: 14,
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildUserCard(User user) {
    return Card(
      elevation: 2,
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
      child: Padding(
        padding: EdgeInsets.all(16),
        child: Row(
          children: [
            CircleAvatar(
              radius: 30,
              backgroundImage: NetworkImage(user.avatar),
            ),
            SizedBox(width: 16),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Row(
                    children: [
                      Text(
                        user.username,
                        style: TextStyle(fontWeight: FontWeight.bold),
                      ),
                      if (user.isVerified)
                        Padding(
                          padding: EdgeInsets.only(left: 4),
                          child: Icon(Icons.verified, color: Colors.blue, size: 16),
                        ),
                    ],
                  ),
                  Text(user.displayName),
                  Text(
                    '${_formatCount(user.followersCount)} followers',
                    style: TextStyle(color: Colors.grey, fontSize: 12),
                  ),
                ],
              ),
            ),
            OutlinedButton(
              onPressed: () => _followUser(user),
              child: Text('Follow'),
              style: OutlinedButton.styleFrom(
                side: BorderSide(color: Theme.of(context).primaryColor),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildPlaceCard(Map<String, String> place) {
    return GestureDetector(
      onTap: () => _searchByPlace(place['name']!),
      child: Card(
        elevation: 2,
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
        child: Container(
          height: 120,
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(12),
            image: DecorationImage(
              image: NetworkImage(place['image']!),
              fit: BoxFit.cover,
            ),
          ),
          child: Container(
            padding: EdgeInsets.all(16),
            decoration: BoxDecoration(
              borderRadius: BorderRadius.circular(12),
              gradient: LinearGradient(
                begin: Alignment.topCenter,
                end: Alignment.bottomCenter,
                colors: [Colors.transparent, Colors.black54],
              ),
            ),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Spacer(),
                Text(
                  place['name']!,
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 18,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                Text(
                  place['posts']!,
                  style: TextStyle(
                    color: Colors.white70,
                    fontSize: 14,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  List<Color> _getTagGradient(String tag) {
    switch (tag.toLowerCase()) {
      case '#trending':
        return [Colors.orange, Colors.red];
      case '#photography':
        return [Colors.blue, Colors.purple];
      case '#travel':
        return [Colors.teal, Colors.blue];
      case '#food':
        return [Colors.orange, Colors.deepOrange];
      case '#fashion':
        return [Colors.pink, Colors.purple];
      case '#fitness':
        return [Colors.green, Colors.teal];
      case '#art':
        return [Colors.purple, Colors.pink];
      case '#nature':
        return [Colors.green, Colors.lightGreen];
      default:
        return [Colors.grey[600]!, Colors.grey[800]!];
    }
  }

  String _formatCount(int count) {
    if (count < 1000) return count.toString();
    if (count < 1000000) return '${(count / 1000).toStringAsFixed(1)}K';
    return '${(count / 1000000).toStringAsFixed(1)}M';
  }

  void _performSearch(String query) {
    if (query.trim().isNotEmpty) {
      Navigator.pushNamed(context, '/search-results', arguments: query);
    }
  }

  void _searchByTag(String tag) {
    Navigator.pushNamed(context, '/search-results', arguments: tag);
  }

  void _searchByPlace(String place) {
    Navigator.pushNamed(context, '/search-results', arguments: place);
  }

  void _viewPost(Post post) {
    Navigator.pushNamed(context, '/post-detail', arguments: post);
  }

  void _followUser(User user) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text('Following ${user.username}'),
        action: SnackBarAction(
          label: 'View Profile',
          onPressed: () {
            Navigator.pushNamed(context, '/user-profile', arguments: user.username);
          },
        ),
      ),
    );
  }

  Future<void> _refreshTrending() async {
    // Simulate API call
    await Future.delayed(Duration(seconds: 2));

    setState(() {
      // Shuffle posts to simulate new trending content
      trendingPosts.shuffle();
    });
  }

  @override
  void dispose() {
    _searchController.dispose();
    _tabController.dispose();
    super.dispose();
  }
}

class Post {
  final String id;
  final String imageUrl;
  final int likes;
  final int comments;
  final String username;
  final bool isVideo;
  final String? duration;
  final List<String> tags;

  Post({
    required this.id,
    required this.imageUrl,
    required this.likes,
    required this.comments,
    required this.username,
    required this.isVideo,
    this.duration,
    required this.tags,
  });
}

class User {
  final String username;
  final String displayName;
  final String avatar;
  final bool isVerified;
  final int followersCount;

  User({
    required this.username,
    required this.displayName,
    required this.avatar,
    required this.isVerified,
    required this.followersCount,
  });
}